File: recamera-intellisense-mcp/src/detection.rs
//! High-level detection facade composing `api::model`, `api::rule`, `api::daemon`,
//! and root `storage::ensure_storage`.

use anyhow::{bail, Result};

use crate::api::{daemon as api_daemon, model as api_model, rule as api_rule};
use crate::api_client::ApiClient;
use crate::storage;
use crate::types::{
    DetectionEvent, DetectionModel, DetectionRule, DeviceRecord, RecordSource, RecordTrigger,
    RuleConfig, ScheduleRange, WriterConfig,
};

// MARK: Models

pub async fn get_detection_models_info(
    client: &ApiClient,
    device: &DeviceRecord,
) -> Result<Vec<DetectionModel>> {
    api_model::list_models(client, device).await
}

pub async fn get_detection_model(
    client: &ApiClient,
    device: &DeviceRecord,
) -> Result<Option<DetectionModel>> {
    api_model::get_active_model(client, device).await
}

pub async fn set_detection_model(
    client: &ApiClient,
    device: &DeviceRecord,
    model_id: Option<i32>,
    model_name: Option<&str>,
) -> Result<()> {
    if model_id.is_none() == model_name.is_none() {
        bail!("Provide exactly one of 'model_id' or 'model_name'.");
    }
    let models = api_model::list_models(client, device).await?;
    let target = if let Some(id) = model_id {
        models
            .iter()
            .find(|m| m.id == id)
            .ok_or_else(|| anyhow::anyhow!("Model ID '{id}' not found on device."))?
    } else {
        let name = model_name.unwrap();
        models
            .iter()
            .find(|m| m.name == name)
            .ok_or_else(|| anyhow::anyhow!("Model name '{name}' not found on device."))?
    };
    api_model::set_active_model(client, device, target).await
}

// MARK: Schedule (detection-level wrapper)

pub async fn get_detection_schedule(
    client: &ApiClient,
    device: &DeviceRecord,
) -> Result<Option<Vec<ScheduleRange>>> {
    api_rule::get_schedule(client, device).await
}

pub async fn set_detection_schedule(
    client: &ApiClient,
    device: &DeviceRecord,
    schedule: Option<&[ScheduleRange]>,
) -> Result<()> {
    api_rule::set_schedule(client, device, schedule).await
}

// MARK: Rules (INFERENCE_SET trigger)

pub async fn get_detection_rules(
    client: &ApiClient,
    device: &DeviceRecord,
) -> Result<Vec<DetectionRule>> {
    let cfg = match api_rule::get_config(client, device).await {
        Ok(c) => c,
        Err(_) => return Ok(vec![]),
    };
    if !is_record_image_enabled(&cfg) {
        return Ok(vec![]);
    }
    // Kind gate on the raw payload: a retired-but-unmigrated SED selection
    // yields [] instead of an error.
    let raw = api_rule::get_record_rule_json(client, device).await?;
    Ok(api_rule::inference_rules_from_raw(&raw))
}

/// Vision intent assumed when a rule omits `source_filter`: an empty filter
/// matches EVERY source — including acoustic classifications — which is
/// almost never what an agent compiling a vision rule wants. Use the raw
/// `set_record_trigger` for an explicit all-sources rule.
const DEFAULT_VISION_SOURCE: &str = "builtin";

pub async fn set_detection_rules(
    client: &ApiClient,
    device: &DeviceRecord,
    rules: &[DetectionRule],
) -> Result<()> {
    for (idx, rule) in rules.iter().enumerate() {
        validate_confidence_range(idx, &rule.confidence_range_filter)?;
    }
    let mut prepared: Vec<DetectionRule> = rules.to_vec();
    for rule in prepared.iter_mut() {
        if rule.source_filter.is_empty() {
            rule.source_filter = vec![DEFAULT_VISION_SOURCE.to_string()];
        }
    }
    if !prepared.is_empty() {
        let sources = api_rule::get_record_sources(client, device).await?;
        validate_rules_against_sources(&prepared, &sources)?;
    }
    ensure_record_image(client, device).await?;
    storage::ensure_storage(client, device).await?;
    let trigger = RecordTrigger::InferenceSet { rules: prepared };
    api_rule::set_trigger(client, device, &trigger).await
}

/// Compile-time check: unknown source ids and unproducible labels fail loudly
/// instead of writing a rule that can never fire. A selected source whose
/// class set is unknowable (the builtin vision source follows the selected
/// model) disables the label check — never cry wolf.
fn validate_rules_against_sources(rules: &[DetectionRule], sources: &[RecordSource]) -> Result<()> {
    for rule in rules {
        let mut picked: Vec<&RecordSource> = Vec::new();
        for sid in &rule.source_filter {
            match sources.iter().find(|s| &s.id == sid) {
                Some(s) => picked.push(s),
                None => bail!(
                    "rule {:?}: unknown source_filter id {:?}; available: {:?}",
                    rule.name,
                    sid,
                    sources.iter().map(|s| s.id.as_str()).collect::<Vec<_>>()
                ),
            }
        }
        if rule.label_filter.is_empty() {
            continue;
        }
        if picked.iter().any(|s| s.kind == "builtin") {
            continue; // builtin classes follow the active vision model: unknowable
        }
        let producible: std::collections::HashSet<&str> = picked
            .iter()
            .flat_map(|s| s.classes.iter().map(|c| c.as_str()))
            .collect();
        let bad: Vec<&str> = rule
            .label_filter
            .iter()
            .map(|l| l.as_str())
            .filter(|l| !producible.contains(l))
            .collect();
        if !bad.is_empty() {
            let states = picked
                .iter()
                .map(|s| {
                    format!(
                        "{}({}, {} classes)",
                        s.id,
                        if s.running { "running" } else { "STOPPED" },
                        s.classes.len()
                    )
                })
                .collect::<Vec<_>>()
                .join(", ");
            bail!(
                "rule {:?}: label(s) {:?} cannot be produced by the selected source(s) [{}]; \
                 check get_record_sources and the AcousticsLab/App Center state first",
                rule.name,
                bad,
                states
            );
        }
    }
    Ok(())
}

fn validate_confidence_range(idx: usize, range: &[f64]) -> Result<()> {
    if range.len() != 2 {
        bail!(
            "rule #{idx}: confidence_range_filter must be exactly [min, max]; got {} value(s)",
            range.len()
        );
    }
    let (min, max) = (range[0], range[1]);
    if !(0.0..=1.0).contains(&min) || !(0.0..=1.0).contains(&max) {
        bail!(
            "rule #{idx}: confidence_range_filter values must be within [0.0, 1.0]; got [{min}, {max}]"
        );
    }
    if min > max {
        bail!("rule #{idx}: confidence_range_filter min ({min}) must be <= max ({max})");
    }
    Ok(())
}

// MARK: Events (daemon)

pub async fn get_detection_events(
    client: &ApiClient,
    device: &DeviceRecord,
    start_unix_ms: Option<i64>,
    end_unix_ms: Option<i64>,
) -> Result<Vec<DetectionEvent>> {
    api_daemon::get_events(client, device, start_unix_ms, end_unix_ms).await
}

pub async fn clear_detection_events(client: &ApiClient, device: &DeviceRecord) -> Result<()> {
    api_daemon::clear_events(client, device).await
}

// MARK: Helpers

fn is_record_image_enabled(cfg: &RuleConfig) -> bool {
    cfg.rule_enabled && cfg.writer.format.eq_ignore_ascii_case("JPG")
}

async fn ensure_record_image(client: &ApiClient, device: &DeviceRecord) -> Result<()> {
    let cfg = api_rule::get_config(client, device).await.ok();
    if cfg.as_ref().is_some_and(is_record_image_enabled) {
        return Ok(());
    }
    let new_cfg = RuleConfig {
        rule_enabled: true,
        writer: WriterConfig {
            format: "JPG".into(),
            interval_ms: 0,
        },
    };
    api_rule::set_config(client, device, &new_cfg).await
}

#[cfg(test)]
mod tests {
    use super::validate_confidence_range;

    #[test]
    fn validate_confidence_range_accepts_valid() {
        assert!(validate_confidence_range(0, &[0.0, 1.0]).is_ok());
        assert!(validate_confidence_range(0, &[0.25, 0.75]).is_ok());
        assert!(validate_confidence_range(0, &[0.5, 0.5]).is_ok());
    }

    #[test]
    fn validate_confidence_range_rejects_bad_length() {
        assert!(validate_confidence_range(0, &[]).is_err());
        assert!(validate_confidence_range(0, &[0.5]).is_err());
        assert!(validate_confidence_range(0, &[0.0, 0.5, 1.0]).is_err());
    }

    #[test]
    fn validate_confidence_range_rejects_out_of_range() {
        assert!(validate_confidence_range(0, &[-0.01, 0.5]).is_err());
        assert!(validate_confidence_range(0, &[0.0, 1.01]).is_err());
    }

    #[test]
    fn validate_confidence_range_rejects_inverted() {
        assert!(validate_confidence_range(0, &[0.8, 0.2]).is_err());
    }
}

File: recamera-intellisense-sdk/recamera_intellisense/detection.py
"""High-level detection facade: schedule + rules + events (wraps ``rule`` and ``files``)."""

from __future__ import annotations

if __name__ == "__main__" and __package__ is None:
    import os
    import sys

    sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    from recamera_intellisense._cli import main

    raise SystemExit(main())

from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from . import files as _files
from . import rule as _rule
from . import storage as _storage
from ._coerce import to_bool

__all__ = [
    "get_detection_schedule",
    "set_detection_schedule",
    "get_detection_rules",
    "set_detection_rules",
    "get_detection_events",
    "clear_detection_events",
]


def get_detection_schedule(device_name: Optional[str] = None) -> Optional[List[Dict[str, str]]]:
    """Alias for :func:`rule.get_schedule_rule`."""
    return _rule.get_schedule_rule(device_name)


def set_detection_schedule(
    device_name: Optional[str] = None,
    schedule: Optional[List[Dict[str, str]]] = None,
) -> None:
    """Alias for :func:`rule.set_schedule_rule`.

    `schedule` defaults to `None` (meaning "always active"), so omitting
    the key from the CLI payload is equivalent to passing `schedule: []`
    or `schedule: null` and clears any existing window.
    """
    _rule.set_schedule_rule(device_name, schedule)


def get_detection_rules(device_name: Optional[str] = None) -> List[Dict[str, Any]]:
    """Active INFERENCE_SET rules, or `[]` when the trigger is not INFERENCE_SET
    (including a retired-but-unmigrated ``SED`` selection)."""
    return list(_rule._get_inference_rules(device_name))


DEFAULT_VISION_SOURCE = "builtin"


def _validate_rules_against_sources(
    rules: List[Dict[str, Any]], sources: List[Dict[str, Any]]
) -> None:
    """Fail loudly on unknown source ids or labels the selected sources can
    never produce — a silently never-firing rule is the worst outcome. The
    ``builtin`` source's classes follow the selected vision model (unknowable
    here), so its presence skips the label check: never cry wolf."""
    by_id = {s["id"]: s for s in sources}
    for rule in rules:
        name = rule.get("name", "")
        selected = list(rule.get("source_filter") or [])
        if not selected:  # explicit empty = all sources
            selected = list(by_id)
        unknown = [sid for sid in selected if sid not in by_id]
        if unknown:
            raise ValueError(
                f"rule {name!r}: unknown source_filter id(s) {unknown}; "
                f"available: {sorted(by_id)}"
            )
        labels = list(rule.get("label_filter") or [])
        if not labels:
            continue
        picked = [by_id[sid] for sid in selected]
        if any(s["kind"] == "builtin" for s in picked):
            continue  # builtin classes follow the active vision model: unknowable
        producible = {c for s in picked for c in s["classes"]}
        bad = [lbl for lbl in labels if lbl not in producible]
        if bad:
            states = ", ".join(
                f"{s['id']}({'running' if s['running'] else 'STOPPED'}, "
                f"{len(s['classes'])} classes)" for s in picked
            )
            raise ValueError(
                f"rule {name!r}: label(s) {bad} cannot be produced by the "
                f"selected source(s) [{states}]; check get_record_sources and "
                f"the AcousticsLab/App Center state before compiling this rule"
            )


def set_detection_rules(
    device_name: Optional[str] = None,
    *,
    rules: List[Dict[str, Any]],
    ensure_writer: bool = True,
    ensure_storage: bool = True,
) -> None:
    """Install an INFERENCE_SET trigger with *rules*.

    Rules without ``source_filter`` are scoped to the ``builtin`` vision
    source — an empty filter matches EVERY source, including acoustic
    classifications. Pass ``source_filter=[\"acousticslab\"]`` for sound.
    Source ids and labels are validated against ``get_record_sources`` first.

    Also (by default):
      * enables the rule pipeline with JPG writer (`ensure_writer=True`);
      * ensures a storage slot is available (`ensure_storage=True`).
    """
    if not isinstance(rules, list):
        raise ValueError("'rules' must be a list of detection-rule dicts.")
    ensure_writer = to_bool(ensure_writer, "ensure_writer")
    ensure_storage = to_bool(ensure_storage, "ensure_storage")
    prepared: List[Dict[str, Any]] = []
    for rule in rules:
        r = dict(rule)
        if r.get("source_filter") is None:
            r["source_filter"] = [DEFAULT_VISION_SOURCE]
        prepared.append(r)
    if prepared:
        _validate_rules_against_sources(prepared, _rule.get_record_sources(device_name))
    if ensure_storage:
        _storage.ensure_storage(device_name)
    trigger = {"kind": "inference_set", "rules": prepared}
    _rule.set_record_trigger(device_name, trigger=trigger)
    if ensure_writer:
        cfg = _rule.get_record_config(device_name)
        needs_update = (
            not cfg["rule_enabled"] or cfg["writer"].get("format", "").upper() != "JPG"
        )
        if needs_update:
            _rule.set_record_config(
                device_name,
                rule_enabled=True,
                writer_format="JPG",
                writer_interval_ms=cfg["writer"].get("interval_ms", 0),
            )


def get_detection_events(
    device_name: Optional[str] = None,
    *,
    start_unix_ms: Optional[int] = None,
    end_unix_ms: Optional[int] = None,
) -> List[Dict[str, Any]]:
    """Normalized detection events, shape-compatible with the MCP server.

    Each event is `{timestamp, timestamp_unix_ms, rule_name, snapshot_path?}` where
    `timestamp` is an ISO-8601 UTC string. Use :func:`files.get_intellisense_events`
    to access the raw daemon payloads instead.
    """
    raw = _files.get_intellisense_events(
        device_name, start_unix_ms=start_unix_ms, end_unix_ms=end_unix_ms
    )
    out: List[Dict[str, Any]] = []
    for item in raw:
        if not isinstance(item, dict):
            continue
        ts = item.get("timestamp")
        if not isinstance(ts, (int, float)):
            continue
        ts_ms = int(ts)
        rule_id = (
            (item.get("id") or "").strip() if isinstance(item.get("id"), str) else ""
        )
        rule_name = rule_id or str(item.get("type", ""))
        file_event = item.get("file_event")
        snapshot_path: Optional[str] = None
        if isinstance(file_event, dict):
            p = file_event.get("path")
            if isinstance(p, str) and p:
                snapshot_path = p
        event = {
            "timestamp": datetime.fromtimestamp(ts_ms / 1000.0, tz=timezone.utc)
            .isoformat()
            .replace("+00:00", "Z"),
            "timestamp_unix_ms": ts_ms,
            "rule_name": rule_name,
        }
        if snapshot_path is not None:
            event["snapshot_path"] = snapshot_path
        out.append(event)
    return out


def clear_detection_events(device_name: Optional[str] = None) -> None:
    """Alias for :func:`files.clear_intellisense_events`."""
    _files.clear_intellisense_events(device_name)


COMMANDS = {
    "get_detection_schedule": get_detection_schedule,
    "set_detection_schedule": set_detection_schedule,
    "get_detection_rules": get_detection_rules,
    "set_detection_rules": set_detection_rules,
    "get_detection_events": get_detection_events,
    "clear_detection_events": clear_detection_events,
}
